rfd/0123-tls-routing-behind-layer7-lb.md

```diff
@@ -228,6 +228,28 @@ See [RFD 0121 - Kubernetes MFA
 sessions](https://github.com/gravitational/teleport/blob/master/rfd/0121-kube-mfa-sessions.md)
 for more details.
 
+### Client source IPs
+
+Correct client source IPs must be reported for IP pinning and audit logs.
+
+Unlike layer 4 load balancers, layer 7 load balancers usually CANNOT transparently forward original clients' IPs or add
+standard [PROXY](https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt) protocol headers to layer 7 payload. The
+standard way for layer 7 load balancers to report correct client source IPs is through the
+["X-Forwarded-For"](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-For) (XFF) header.
+
+The Proxy service will examine the values of the XFF headers when `proxy_service.use_xff_header` is enabled (more
+details in security section).
+
+ALL web APIs will take client source IPs from the XFF headers. This includes login APIs, Web UI APIs (including wss
+sessions) and the ALPN connection upgrade API, to make sure a correct IP is always logged.
+
+Teleport should support XFF header with both IPv4 and IPv6 addresses, with or without port numbers. In case no port is
+reported in the XFF header, port can be taken from the observed TCP connection or use an arbitrary number, since the
+port is usually trivial.
+
+To keep things simple, only XFF header containing one IP address is accepted. Any requests with multiple XFF headers or
+headers containing multiple IPs will be rejected.
```

Comment:
I think this is sensible from a time-saving standpoint, but I suspect it will cause problems for people running in Kubernetes behind ALBs and using instance-based target groups, where the traffic will go from client -> ALB -> node `kube-proxy` -> pod and therefore probably have two `X-Forwarded-For` headers set.

If we're good with this for a first pass that's OK, but I suspect people using this in anger will have more complex traffic routing setups.

Reply:
I think `kube-proxy` is usually layer 4 and shouldn't overwrite XFF headers? Ingress controllers can be set to preserve XFF header from ALB I believe.

Reply:
kube-proxy is L4 and won't affect the connection headers because it just forwards.
MetalLb can run in L2 or BGP mode so it won't affect the request headers.

Reply:
Cool, my mistake then.